### PR TITLE
Combined Crystal environment

### PR DESCRIPTION
**Pull request by michalkoziarski**
Re-opened PR because I had to rename the branch to avoid confusion.

- Implements Crystal environment, combining Composition, SpaceGroup and LatticeParameters in a sequential order.
- Adds QoL change to LatticeParameters to accept arbitrary grid size.
- Changes Composition oracle state format to contain atom counts (instead of Li atom counts, total atom counts, and frequencies as before).
- Contains some minor fixes to Composition.
- (Temporarily) disables Composition's neutral charge check - which would cause issues with Crystal environment getting stuck occasionally, and would significantly reduce training speed of GFlowNet. This needs to be rewritten to be included in forward mask.

### CODE REVIEW

File: config/env/crystal.yaml
Comment (alexhernandezgarcia) [inline config/env/crystal.yaml:1]: Why not having this inside directory `crystals/`?

File: gflownet/envs/crystals/composition.py
Comment (alexhernandezgarcia) [inline gflownet/envs/crystals/composition.py:355]: How replacing the commented out piece of code by
```
if self.do_charge_check:
  # do charge check
```
and `do_charge_check` is control by config?

File: gflownet/envs/crystals/crystal.py
Comment (alexhernandezgarcia) [inline gflownet/envs/crystals/crystal.py:24]: Sorry, but why these values? Could we add some more info to the docstring?

File: gflownet/envs/crystals/crystal.py
Comment (alexhernandezgarcia) [inline gflownet/envs/crystals/crystal.py:62]: states (and source) are lists in all other environments. Is it here a tuple for a particular reason?

File: gflownet/envs/crystals/crystal.py
Comment (alexhernandezgarcia) [inline gflownet/envs/crystals/crystal.py:92]:
```suggestion
        # EOS is EOS of LatticeParameters because it is the last stage
        self.eos = self.lattice_parameters.eos
```

File: gflownet/envs/crystals/crystal.py
Comment (alexhernandezgarcia) [inline gflownet/envs/crystals/crystal.py:99]:
```suggestion
```
Unnecessary because set in `super().__init__(**kwargs)`

File: gflownet/envs/crystals/crystal.py
Comment (alexhernandezgarcia) [inline gflownet/envs/crystals/crystal.py:117]: I think this is incorrect for a couple of reasons:
- 1. get_crystal_system is meant to return the name of the crystal system. Actually, it will never return None but instead it will return the string "None" if the crystal system has not been set. 
- 2. I think what we want here is anyway not the crystal system but the crystal lattice system, don't we? That would be:
```suggestion
        if self.space_group.state[0] == 0:
            raise ValueError(
                "Cannot set lattice parameters without crystal system determined in the space group."
            )
```

File: gflownet/envs/crystals/crystal.py
Comment (alexhernandezgarcia) [inline gflownet/envs/crystals/crystal.py:122]: Shouldn't the lattice parameters work with crystal-lattice system instead of crystal system?

File: gflownet/envs/crystals/crystal.py
Comment (alexhernandezgarcia) [inline gflownet/envs/crystals/crystal.py:130]: The multiplication thing is probably fine because it is a tuple but flagging this just in case it could have some unwanted side effects. @vict0rsch ?

File: gflownet/envs/crystals/crystal.py
Comment (alexhernandezgarcia) [inline gflownet/envs/crystals/crystal.py:147]: These for loops can be avoided by storing the max values, right? This may not be negligible because the action spaces are large and this will be called many many times (each call to `step()` for example)

Comment (alexhernandezgarcia) [reply gflownet/envs/crystals/crystal.py:147]: Actually, now I think about it, the length of the actions should be the same for a given environment. At least this is the case for the space group environment. I started coding things this way since I decided that at some point I want to port this to all-torch-tensors.

File: gflownet/envs/crystals/crystal.py
Comment (alexhernandezgarcia) [inline gflownet/envs/crystals/crystal.py:170]: Nice check! :)

File: gflownet/envs/crystals/crystal.py
Comment (alexhernandezgarcia) [inline gflownet/envs/crystals/crystal.py:50]: This is a bit annoying no? How about we make `LatticeParameters` be unconstrained by default? Would that solve the mentioned stateless issues with the lattice parameters?

File: gflownet/envs/crystals/crystal.py
Comment (alexhernandezgarcia) [inline gflownet/envs/crystals/crystal.py:267]: I am not sure I understand what you mean by this? Mind elaborate?

File: gflownet/envs/crystals/crystal.py
Comment (alexhernandezgarcia) [inline gflownet/envs/crystals/crystal.py:283]: Should this not be a method of `Stage` class instead?

File: gflownet/envs/crystals/crystal.py
Comment (alexhernandezgarcia) [inline gflownet/envs/crystals/crystal.py:288]: Should this not be a method of `Stage` class instead?

File: gflownet/envs/crystals/crystal.py
Comment (alexhernandezgarcia) [inline gflownet/envs/crystals/crystal.py:366]: How about making creating a method `_build_state(stage, composition, space_group, lattice_parameters)`?

File: gflownet/envs/crystals/crystal.py
Comment (alexhernandezgarcia) [inline gflownet/envs/crystals/crystal.py:403]:
```suggestion
        Prepares a list of states in "GFlowNet format" for the oracle. Simply
        a concatenation of all crystal components.
```

File: config/env/crystal.yaml
Comment (alexhernandezgarcia) [inline config/env/crystal.yaml:15]:
```suggestion
  grid_size: 10
```

File: gflownet/envs/crystals/lattice_parameters.py
Comment (alexhernandezgarcia) [inline gflownet/envs/crystals/lattice_parameters.py:44]:
```suggestion
        grid_size: int = 10,
```

**Review by alexhernandezgarcia — CHANGES_REQUESTED** [review]
Really nice work! I generally looks good to me. I have made some suggestions that I would like to have or discuss before merging, but nothing is critical.

File: gflownet/envs/crystals/crystal.py
Comment (michalkoziarski) [reply gflownet/envs/crystals/crystal.py:62]: It is a list :) parenthesis are just to group the summation operation.

File: gflownet/envs/crystals/crystal.py
Comment (michalkoziarski) [reply gflownet/envs/crystals/crystal.py]: It's a tuple of ints that is never modified afterwards, I believe it should be find.

File: gflownet/envs/crystals/crystal.py
Comment (michalkoziarski) [reply gflownet/envs/crystals/crystal.py]: Good point about computational cost, I changed it to be computed only once.

Regarding same action length, I know, and I believe this is the case right now, but I think it's worth checking it once anyway to prevent issues with any possible future changes.

File: gflownet/envs/crystals/crystal.py
Comment (michalkoziarski) [reply gflownet/envs/crystals/crystal.py:67]: Stateless thing is a separate issue that wouldn't be solved by this. Yes, it is annoying, but I don't think making `LatticeParameters` unconstrained will help - the thing is that `LatticeParameters` just has to have constraints to work properly (to determine the source, masks etc), and if anything, removing the need for the constraint will make it work like using the triclinic system anyway, while making the `LatticeParameters` have additional edge case to consider internally.

File: gflownet/envs/crystals/crystal.py
Comment (michalkoziarski) [reply gflownet/envs/crystals/crystal.py:151]: I am not sure what you mean? `space_group.lattice_system` is being used currently in the code you cited, so that's not the crystal system, right?

Comment (michalkoziarski) [reply gflownet/envs/crystals/crystal.py:149]: (I think the confusion might have been, I was checking for `crystal_system` being set and raised error if it wasn't, but the `LatticeParameters` were, I think properly, conditioned on the `lattice_system`)

File: gflownet/envs/crystals/crystal.py
Comment (alexhernandezgarcia) [reply gflownet/envs/crystals/crystal.py]: 2. No, the lattice system and the crystal system are set simultaneously because they are both set via the crystal-lattice system, which is the first dimension of the space group env.

Comment (alexhernandezgarcia) [reply gflownet/envs/crystals/crystal.py]: Note that I have made an edit of the error message too

**Comment by alexhernandezgarcia** [comment]
Thanks for the changes and answers! This can be merged for me - pending the related PR.

**Pull request merged**
